trace_decoder/src/protocol_processing.rs

```diff
@@ -0,0 +1,173 @@
+    Smt(SmtProcessedBlockTracePreImages),
+}
+
+/// Processes any type of pre-image
+pub(crate) fn process_block_trace_trie_pre_images(
+    image: TriePreImage,
+) -> TraceParsingResult<Box<ProcessedPreImage>> {
+    Ok(Box::new(match image {
+        TriePreImage::Mpt(images) => ProcessedPreImage::Mpt(process_mpt_trie_images(images)?),
+        TriePreImage::Smt(images) => ProcessedPreImage::Smt(process_smt_trie_images(images)?),
+    }))
+}
+
+/// Process a block trace pre-image compact into an MPT payload. Will return an
+/// error if given an SMT payload.
+pub(crate) fn process_mpt_block_trace_trie_pre_images(
+    image: TriePreImage,
+) -> TraceParsingResult<MptProcessedBlockTracePreImages> {
+    let res = process_mpt_trie_images(
+        image
+            .into_mpt()
+            .map_err(|_| TraceParsingError::UnexpectedTraceFormat)?,
+    )?;
+
+    Ok(res)
+}
+
+/// Process a block trace pre-image compact into an SMT payload. Will return an
+/// error if given an MPT payload.
+pub(crate) fn process_smt_block_trace_trie_pre_images(
+    image: TriePreImage,
+) -> TraceParsingResult<SmtProcessedBlockTracePreImages> {
+    let res = process_smt_trie_images(
+        image
+            .into_smt()
+            .map_err(|_| TraceParsingError::UnexpectedTraceFormat)?,
+    )?;
+
+    Ok(res)
+}
+
+pub(crate) fn process_mpt_trie_images(
+    images: MptBlockTraceTriePreImages,
+) -> MptTraceParsingResult<MptProcessedBlockTracePreImages> {
+    match images {
+        MptBlockTraceTriePreImages::Separate(t) => process_separate_trie_pre_images(t),
+        MptBlockTraceTriePreImages::Combined(t) => process_combined_trie_pre_images(t),
+    }
+}
+
+fn process_combined_trie_pre_images(
+    tries: MptCombinedPreImages,
+) -> MptTraceParsingResult<MptProcessedBlockTracePreImages> {
+    process_compact_trie(tries.compact)
+}
+
+fn process_separate_trie_pre_images(
+    tries: MptSeparateTriePreImages,
+) -> MptTraceParsingResult<MptProcessedBlockTracePreImages> {
+    let tries = MptPartialTriePreImages {
+        state: process_state_trie(tries.state),
+        storage: process_storage_tries(tries.storage),
+    };
+
+    Ok(MptProcessedBlockTracePreImages {
+        tries,
+        extra_code_hash_mappings: None,
+    })
+}
+
+fn process_state_trie(trie: MptSeparateTriePreImage) -> HashedPartialTrie {
+    match trie {
+        MptSeparateTriePreImage::Uncompressed(_) => todo!(),
+        MptSeparateTriePreImage::Direct(t) => t.0,
+    }
+}
+
+fn process_storage_tries(
+    trie: MptSeparateStorageTriesPreImage,
+) -> HashMap<HashedAccountAddr, HashedPartialTrie> {
+    match trie {
+        MptSeparateStorageTriesPreImage::SingleTrie(t) => process_single_combined_storage_tries(t),
+        MptSeparateStorageTriesPreImage::MultipleTries(t) => process_multiple_storage_tries(t),
+    }
+}
+
+fn process_single_combined_storage_tries(
+    _trie: MptTrieUncompressed,
+) -> HashMap<HashedAccountAddr, HashedPartialTrie> {
+    todo!()
```

Comment:
For the upcoming pr? Okay there's hundreds of these so I'm going to give up writing this and let's pretend this is pointing at all the `todo!()`

Reply:
Yeah. So more-or-less, this PR just gets a lot of the structural changes done. The next PR integrates Arpit's work, and the third and final one fills everything else in.